code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -605,28 +626,33 @@ runBroker = do
              A.withAsync (clientsToBrokerWorker cEnv) $ const $
              action
 
-    liftIO $ withWorkers $ do
-      let action = do
-              cpd@CliPollData{..} <- atomically $ readTVar ztCliPollData
-
-              events <- Z.poll (-1) cpdPolls
-              forM_ (events `zip` [0..]) $ \(e,i) ->
-                  unless (null e) $ case resolveSocketIndex i cpd of
-                      BackendSocket (peerId, backSock) -> receiveBack backSock peerId
-                      SubSocket (peerId, subSock)      -> receiveSub subSock peerId
-                      CliSocket -> do
-                          req <- iqReceive ztClientsQueue
-                          whenJust req $ uncurry clientToBackend
-                      ReqSocket -> do
-                          req <- iqReceive ztCliRequestQueue
-                          whenJust req processReq
-
-      forever $
-          (forever action)
-          `catchAny`
-          (\e -> do ztCliLog Error $
-                        "Client broker exited, restarting in 2s: " <> show e
-                    threadDelay 2000000)
+    let pollAndProcess = do
+            cpd@CliPollData{..} <- atomically $ readTVar ztCliPollData
+            events <- Z.poll (-1) cpdPolls
+            forM_ (events `zip` [0..]) $ \(e,i) ->
+                unless (null e) $ case resolveSocketIndex i cpd of
+                    BackendSocket (peerId, backSock) -> receiveBack backSock peerId
+                    SubSocket (peerId, subSock)      -> receiveSub subSock peerId
+                    CliSocket -> do
+                        req <- iqReceive ztClientsQueue
+                        whenJust req $ uncurry clientToBackend
+                    ReqSocket -> do
+                        req <- iqReceive ztCliRequestQueue
+                        whenJust req processReq
+
+    let runAll = do
+            x <- tryTakeMVar ztCliTerminated
+            whenNothing x $
+                error $ "Couldn't start client broker: " <>
+                        "it's either already running, or not initialised"
+            withWorkers $ forever $
+                (forever pollAndProcess)
+                `catchAny`
+                (\e -> do ztCliLog Error $
+                              "Client broker exited, restarting in 2s: " <> show e
+                          threadDelay 2000000)
+
+    liftIO $ runAll `finally` (tryPutMVar ztCliTerminated ())
```

Comment:
This seems to be not entirely safe. If you get thrown an async exception during/upon `tryTakeMVar` call, the node will hang.
(Personally, I find `TVar` with its `retry` simpler/safer for use as a semaphore, but that's a matter of choice.)